### PR TITLE
[AXC-3403] Improve NFC scanning loading bar

### PR DESCRIPTION
**Pull request by mikahautz-mohemian**
[AXC-3403](https://airside.atlassian.net/browse/AXC-3403)

This PR makes changes in what is tracked, and how it is tracked. Before, it was just tracked whenever a data group was fully read. However, since the data groups had drastically different sizes (~90 bytes to ~18kb), the loading bar was very chunky and not continuous at all.

Therefore, the following changes are made:
- Track the read byte size instead of data groups
- Track the total size of each individual data group
- Add tracking for:
      - .COM file read
      - BAC authentication
      - PACE authentication
 
Related to https://github.com/airsidemobile/nfc-sdk-ios/pull/142 and https://github.com/airsidemobile/habicht-ios/pull/2281.

[AXC-3403]: https://airside.atlassian.net/browse/AXC-3403?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Sources/NFCPassportReader/PACEHandler.swift [hunk 1]
```diff
@@ -61,14 +62,15 @@ public class PACEHandler {
     private var digestAlg : String = ""
     private var keyLength : Int = -1
     
-    public init(cardAccess : CardAccess, tagReader: TagReader) throws {
+    public init(cardAccess : CardAccess, tagReader: TagReader, updateMessageHandler : ((NFCViewDisplayMessage) -> Void?)?) throws {
```
Comment (daniel-moh) [inline Sources/NFCPassportReader/PACEHandler.swift:65]: I'm curious about the Void-Optional. 😄 Never seen that before. Why do we need it?

File: Sources/NFCPassportReader/PACEHandler.swift [hunk 2]
```diff
@@ -171,6 +173,7 @@ public class PACEHandler {
     /// - Parameters:
     ///   - passportNonce: The decrypted nonce received from the passport
     func doStep2( passportNonce: [UInt8]) async throws -> OpaquePointer {
+        self.updateMessageHandler?(NFCViewDisplayMessage.paceAuthenticationPercentage(40))
```
Comment (daniel-moh) [inline Sources/NFCPassportReader/PACEHandler.swift:176]: Can you elaborate a bit on that 40 and the 80 below? 🙏

File: Sources/NFCPassportReader/DataGroups/DataGroupId.swift [hunk 3]
```diff
@@ -101,5 +101,29 @@ public enum DataGroupId : Int, CaseIterable {
             case .Unknown:  return nil
         }
     }
+    
+    static func getIDFromTag(_ tag: [UInt8]) -> DataGroupId {
+        switch tag {
+            case [0x01, 0x1E]: return .COM
```
[the remark this answers is not in the document]
Comment (mikahautz-mohemian) [reply Sources/NFCPassportReader/DataGroups/DataGroupId.swift:107]: Good idea! This will make it more maintainable, I'll change it 👍

File: Sources/NFCPassportReader/PACEHandler.swift (hunk 1 above)
Comment (mikahautz-mohemian) [reply Sources/NFCPassportReader/PACEHandler.swift:65]: Good catch, actually we don't need it! 😄 
I think I just copied the closure from here: https://github.com/airsidemobile/NFCPassportReader/blob/4ddd4b2a3e161d48f1c268e91458ecb3a47b31ac/Sources/NFCPassportReader/PassportReader.swift#L69 
Then I changed the return type from String to Void and forgot the delete the Optional. I'll change it 👍

File: Sources/NFCPassportReader/NFCViewDisplayMessage.swift [hunk 4]
```diff
@@ -10,10 +10,12 @@ import Foundation
 @available(iOS 13, macOS 10.15, *)
 public enum NFCViewDisplayMessage {
     case requestPresentPassport
-    case authenticatingWithPassport(Int)
-    case readingDataGroupProgress(DataGroupId, Int)
+    case authenticatingWithPassport
```
[the remark this answers is not in the document]
Comment (mikahautz-mohemian) [reply Sources/NFCPassportReader/NFCViewDisplayMessage.swift:13]: Good point! We could keep the `(Int)` to keep it as it was, even though we don't specifically need it. However, I think there might be more breaking changes, which we do unfortunately need. E.g., adding the `trackingDelegate` to the `TagReader`: https://github.com/airsidemobile/NFCPassportReader/blob/4ddd4b2a3e161d48f1c268e91458ecb3a47b31ac/Sources/NFCPassportReader/TagReader.swift#L26 
or the `updateMessageHandler` to the `PACEHandler`:
https://github.com/airsidemobile/NFCPassportReader/blob/4ddd4b2a3e161d48f1c268e91458ecb3a47b31ac/Sources/NFCPassportReader/PACEHandler.swift#L65

File: Sources/NFCPassportReader/PACEHandler.swift (hunk 2 above)
Comment (mikahautz-mohemian) [reply Sources/NFCPassportReader/PACEHandler.swift:176]: The PACE authentication is split into a 4-step process. It takes quite some time to complete the overall authentication, that's why I split it roughly 40-40-20, meaning step 1 takes approx. 40% of the overall PACE authentication duration, step 2 takes roughly the same amount of time (therefore, 80% of the authentication is complete by then), and steps 3 and 4 combined make up the remaining 20% of the time. 
Basically, these intermediate steps are just added to make the progress bar even smoother. If we only updated the progress bar before PACE starts and after it ends, it would take a bit too long to advance to the next progress dot.

File: Sources/NFCPassportReader/NFCViewDisplayMessage.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (daniel-moh) [reply Sources/NFCPassportReader/NFCViewDisplayMessage.swift:13]: Yeah but both of these are optional parameters and can have a default value of nil to be non-breaking.

File: Sources/NFCPassportReader/PACEHandler.swift (hunk 2 above)
Comment (daniel-moh) [reply Sources/NFCPassportReader/PACEHandler.swift:176]: Sounds good.

File: Sources/NFCPassportReader/NFCViewDisplayMessage.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikahautz-mohemian) [reply Sources/NFCPassportReader/NFCViewDisplayMessage.swift]: You're right. I made changes that add the `(Int)` again, and set default values of `nil` to the optional parameters. LMK what you think! 😄

**Review by daniel-moh — APPROVED** [review]
